webapp-django/crashstats/cron/management/commands/cron.py

```diff
@@ -0,0 +1,464 @@
+
+
+def convert_frequency(value):
+    """return the number of seconds that a certain frequency string represents.
+    For example: `1d` means 1 day which means 60 * 60 * 24 seconds.
+    The recognized formats are:
+        10d  : 10 days
+        3m   : 3 minutes
+        12h  : 12 hours
+    """
+    match = FREQUENCY_RE.match(value)
+    if not match:
+        raise FrequencyDefinitionError(value)
+
+    number = int(match.group(1))
+    unit = match.group(2)
+
+    if unit == 'h':
+        number *= 60 * 60
+    elif unit == 'm':
+        number *= 60
+    elif unit == 'd':
+        number *= 60 * 60 * 24
+    elif unit:
+        raise FrequencyDefinitionError(value)
+    return number
+
+
+VALID_TIME_RE = re.compile(r'^\d\d:\d\d$')
+
+
+def check_time(value):
+    """Validate it's an appropriate time in HH:MM form."""
+    if not VALID_TIME_RE.match(value):
+        raise TimeDefinitionError("Invalid definition of time %r" % value)
+
+    hh, mm = value.split(':')
+    if int(hh) > 23 or int(mm) > 59:
+        raise TimeDefinitionError("Invalid definition of time %r" % value)
+
+
+def get_matching_job_specs(cmds):
+    """Return list of matching job specs for this cmd.
+
+    :arg cmds: "all" returns list of all job_specs, cmds return single job_spec,
+        list of cmds returns list of job_specs
+
+    """
+    if cmds == ['all']:
+        return JOBS
+    if isinstance(cmds, str):
+        return JOBS_MAP.get(cmds)
+    return [JOBS_MAP[cmd] for cmd in cmds if cmd in JOBS_MAP]
+
+
+def time_to_run(job_spec, job):
+    """Determine whether it's time for this cmd to run."""
+    now = timezone.now()
+    time_ = job_spec.get('time')
+
+    if job.next_run is None:
+        if time_:
+            # Only run if this hour and minute is < now
+            h, m = [int(x) for x in time_.split(':')]
+            if now.hour > h:
+                return True
+            elif now.hour == h and now.minute >= m:
+                return True
+            return False
+
+        else:
+            return True
+
+    return job.next_run < now
+
+
+def get_run_times(job_spec, last_success):
+    """Return generator of run_times for this job."""
+    now = timezone.now()
+
+    # If this is not a backfill job, just run it
+    if not job_spec.get('backfill', False):
+        yield now
+        return
+
+    # If this is a backfill command that's never been run before, run it
+    if last_success is None:
+        yield now
+        return
+
+    # Figure out the backfill times and yield those; base it on the
+    # first_run datetime so the job doesn't drift
+    when = last_success
+    if job_spec.get('time'):
+        # So, reset the hour/minute part to always match the
+        # intention.
+        h, m = [int(x) for x in job_spec['time'].split(':')]
+        when = when.replace(
+            hour=h,
+            minute=m,
+            second=0,
+            microsecond=0
+        )
+    seconds = convert_frequency(job_spec.get('frequency'))
```

Comment:
If `frequency` is `None`, then `convert_frequency` will raise an error and this will fail. We should probably have a default frequency of `1d` or something like that.